src/internal/storage/fileset/reader.go

```diff
@@ -27,13 +28,11 @@ func newReader(store MetadataStore, chunks *chunk.Storage, idxCache *index.Cache
 }
 
 func (r *Reader) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
-	prim, err := r.getPrimitive(ctx)
-	if err != nil {
-		return err
-	}
-	ir := index.NewReader(r.chunks, r.idxCache, prim.Additive, opts...)
-	return ir.Iterate(ctx, func(idx *index.Index) error {
-		return cb(newFileReader(r.chunks, idx))
+	return r.withPrimitive(ctx, func(ctx context.Context, prim *Primitive) error {
```

Comment:
I would prefer this to be refactored such that we don't need this method. We should be able to structure it such that we create the subcontext with a utility method, get the primitive fileset with `getPrimitive`, then do whatever afterwards. This method that combines both doesn't seem like the right approach.